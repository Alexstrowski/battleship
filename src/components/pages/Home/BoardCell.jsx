import { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { SVG } from 'assets/svg';
import { updateSunkStatus } from 'utils/gameboard';

const BoardCell = ({
    row,
    column,
    isShip,
    isGameStarted,
    setTurnCounter,
    ship,
    ships,
    setShips,
    isRecord,
    shotsMissed,
    setShotsMissed,
}) => {
    const [selected, isSelected] = useState(
        isRecord && shotsMissed.some((ship) => ship.x === row && ship.y === column) ? true : false,
    );
    const [isSunken, setIsSunken] = useState(false);

    useEffect(() => {
        setIsSunken(ship ? ship.isSunken : false);
    }, [ship]);

    const onClickCell = () => {
        if (!isGameStarted) return;

        if (!selected) {
            setTurnCounter((prevTurn) => prevTurn - 1);

            if (isShip) {
                const id = ships.find((ship) => ship.x === row && ship.y === column).id;
                const updateShips = ships.map((ship) => {
                    return ship.x === row && ship.y === column ? { ...ship, isHit: true } : ship;
                });
                setShips(updateSunkStatus(updateShips, id));
            } else {
                setShotsMissed([...shotsMissed, { x: row, y: column }]);
            }
        }
        isSelected(true);
    };

    return (
        <div
            className="md:h-20 md:w-20 h-10 w-10 md:p-4 bg-white border-white border-2 mr-1 mb-1 rounded-md cursor-pointer"
            onClick={onClickCell}
        >
            {!isRecord ? (
                isSunken ? (
                    <div className="w-full h-full">
                        <img src={SVG[`sunken_ship_${ship.type}`]}></img>
                    </div>
                ) : selected && isShip ? (
                    <div className="w-full h-full">
                        <img src={SVG.explosion}></img>
                    </div>
                ) : selected && !isShip ? (
                    <div className="bg-gray-300 w-full h-full"></div>
                ) : (
                    ''
                )
            ) : ship && ship.isSunken ? (
                <div className="w-full h-full">
                    <img src={SVG[`sunken_ship_${ship.type}`]}></img>
                </div>
            ) : ship && ship.isHit ? (
                <div className="w-full h-full">
                    <img src={SVG.explosion}></img>
                </div>
            ) : selected && !isShip ? (
                <div className="bg-gray-300 w-full h-full"></div>
            ) : (
                ''
            )}
        </div>
    );
};

export default BoardCell;

BoardCell.propTypes = {
    row: PropTypes.number.isRequired,
    column: PropTypes.number.isRequired,
    isShip: PropTypes.bool.isRequired,
    isGameStarted: PropTypes.bool.isRequired,
    setTurnCounter: PropTypes.func,
    ship: PropTypes.object,
    ships: PropTypes.any,
    setShips: PropTypes.func,
    isRecord: PropTypes.bool.isRequired,
    shotsMissed: PropTypes.array,
    setShotsMissed: PropTypes.func,
};
