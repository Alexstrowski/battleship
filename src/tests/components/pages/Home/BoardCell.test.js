import { mount, shallow } from 'enzyme';
import BoardCell from 'components/pages/Home/BoardCell';

describe('Tests in <BoardCell />', () => {
    let wrapper;
    const setTurnCounter = jest.fn();
    const ship = { x: 0, y: 0, isHit: true, isSunken: false, type: 3, id: 'f0ffc56d-077a-4c26-9e27-6f0909288370' };
    const ships = [
        {
            x: 0,
            y: 0,
            isHit: false,
            isSunken: false,
            type: 3,
            id: 'f0ffc56d-077a-4c26-9e27-6f0909288370',
        },
        {
            x: 0,
            y: 1,
            isHit: true,
            isSunken: false,
            type: 3,
            id: 'f0ffc56d-077a-4c26-9e27-6f0909288370',
        },
        {
            x: 0,
            y: 2,
            isHit: true,
            isSunken: false,
            type: 3,
            id: 'f0ffc56d-077a-4c26-9e27-6f0909288370',
        },
    ];
    const setShips = jest.fn();
    const shotsMissed = [];
    const setShotsMissed = jest.fn();

    beforeEach(() => {
        wrapper = shallow(
            <BoardCell
                row={0}
                column={0}
                isShip
                isGameStarted
                setTurnCounter={setTurnCounter}
                ship={ship}
                ships={ships}
                setShips={setShips}
                isRecord={false}
                shotsMissed={shotsMissed}
                setShotsMissed={setShotsMissed}
            />,
        );
    });

    test('Must render <BoardCell /> correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Must render empty cell correctly', () => {
        expect(wrapper.contains('img')).toBe(false);
    });

    test('Must render gray cell correctly', () => {
        const wrapperGrayCell = shallow(
            <BoardCell
                row={0}
                column={0}
                isShip={false}
                isGameStarted
                setTurnCounter={setTurnCounter}
                ship={ship}
                ships={ships}
                setShips={setShips}
                isRecord={false}
                shotsMissed={shotsMissed}
                setShotsMissed={setShotsMissed}
            />,
        );
        wrapperGrayCell.simulate('click');
        const divCell = wrapperGrayCell.find('div').at(1);
        const className = divCell.prop('className');
        expect(className.includes('bg-gray-300')).toBe(true);
    });

    test('Must render explosion cell correctly', () => {
        wrapper.simulate('click');
        const img = wrapper.find('img').at(0);
        expect(img.prop('src')).toBe('explosion.svg');
    });

    test('Must render explosion cell correctly', () => {
        wrapper.simulate('click');
        const img = wrapper.find('img').at(0);
        expect(img.prop('src')).toBe('explosion.svg');
    });

    test('Must render sunken ship cell correctly', () => {
        const sunkenShip = {
            x: 0,
            y: 0,
            isHit: true,
            isSunken: true,
            type: 1,
            id: 'f0ffc56d-077a-4c26-9e27-6f0909288370',
        };
        const sunkenShips = [
            {
                x: 0,
                y: 0,
                isHit: true,
                isSunken: true,
                type: 1,
                id: 'f0ffc56d-077a-4c26-9e27-6f0909288370',
            },
        ];
        const wrapperSunkenShipCell = mount(
            <BoardCell
                row={0}
                column={0}
                isShip
                isGameStarted
                setTurnCounter={setTurnCounter}
                ship={sunkenShip}
                ships={sunkenShips}
                setShips={setShips}
                isRecord={false}
                shotsMissed={shotsMissed}
                setShotsMissed={setShotsMissed}
            />,
        );
        wrapperSunkenShipCell.simulate('click');
        const img = wrapperSunkenShipCell.find('img').at(0);
        expect(img.prop('src')).toBe('ship1.svg');
    });
});
