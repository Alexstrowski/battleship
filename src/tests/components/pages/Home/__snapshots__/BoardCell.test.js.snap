// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <BoardCell /> Must render <BoardCell /> correctly 1`] = `
<div
  className="md:h-20 md:w-20 h-10 w-10 md:p-4 bg-white border-white border-2 mr-1 mb-1 rounded-md cursor-pointer"
  id="board-cell"
  onClick={[Function]}
/>
`;
